### ISSUE TITLE
Unable to build images in Kaniko after update to 0.12.22 (Kaniko + in cluster registry)

### ISSUE THREAD

**Issue by Chipcius** [garden-io/garden#2418]
## Bug
I can no longer build my images in Kaniko after updating to 0.12.22

### Current Behavior

![image](https://user-images.githubusercontent.com/1033039/120078680-26038f80-c0a0-11eb-84b9-6ac8ec9a262d.png)

### Expected behavior

It should build just fine

### Reproducible example

Is easily repeatable in my repository but according to @edvald  it is not easily reproducible on his side.

To make sure I jumped up and down a version a couple of times and completely removed garden from my cluster and re-installed for both versions.

Works fine on 0.12.21 and fails on 0.12.22

Garden files included in additional context

### Workaround

Downgrade back to 0.12.21

### Additional context

project.garden.yaml
```
kind: Project
name: system

environments:
  - name: dev
    defaultNamespace: 'u-${local.env.USER || local.username}'

providers:
  - name: kubernetes
    context: gd1
    #buildMode: cluster-buildkit
    buildMode: kaniko
    namespace:
      name: 'system-u-${local.env.USER || local.username}'
      labels:
        istio-injection: enabled
      annotations:
        field.cattle.io/projectId: "${local.env.GARDEN_RANCHER_PROJECT || '<redacted>'}"
    environments:
      - dev
    storage:
      registry:
        size: 200000
      sync:
        size: 200000


defaultEnvironment: 'dev'

```

module.garden.yaml
```
kind: Module
type: exec
name: yarn-build
build:
  command:
    - "./yarn install && ./yarn build"
include:
  - src/**/*.js
  - src/**/*.jsx
  - src/**/*.ts
  - src/**/*.tsx
  - src/**/*.css
  - src/**/*.scss
  - src/**/*.graphql
  - yarn
  - yarn.lock
  - package.json
  - apollo.config.js
  - babel.config.js
  - .eslintrc.json
  - inner_modules
  - jest.config.js
  - postcss.config.js
  - .prettierrc
  - tsconfig.json
  - webpack.config.js
  - webpack.dev.js
  - public/inc
---
kind: Module
type: container
name: system-fpm
dockerfile: Dockerfile
build:
  dependencies:
  - name: yarn-build
    copy:
      - source: public/bundles
        target: public/bundles
hotReload:
  sync:
  - source: src
    target: /var/www/src
  - source: vendor
    target: /var/www/vendor
services:
  - name: system-fpm
    ports:
      - name: fpm
        containerPort: 9000
        servicePort: 9000
---
kind: Module
type: container
name: system
dockerfile: Dockerfile.nginx
build:
  dependencies:
  - name: yarn-build
    copy:
      - source: public/bundles
        target: public/bundles
services:
  - name: system
    ports:
      - name: http
        containerPort: 80
        servicePort: 80
---
kind: Module
type: kubernetes
name: system-gateway
manifests:
  - apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: system-gwy
    spec:
      selector:
        istio: ingressgateway # use istio default controller
      servers:
      - port:
          number: 80
          name: http
          protocol: HTTP
        hosts:
        - ${local.env.USER || local.username}.<redacted>
  - apiVersion: networking.istio.io/v1alpha3
    kind: VirtualService
    metadata:
      name: virtual-svc
    spec:
      hosts:
      - ${local.env.USER || local.username}.<redacted>
      gateways:
      - system-gwy
      http:
      - route:
        - destination:
            host: system
```

### Your environment

* **OS:** Ubuntu 20.04
* **How I'm running Kubernetes:** GKE
* **Garden version:** 0.12.22

**Comment by eysi09** [comment]
Hi @Chipcius and thanks for creating the issue. 

We're looking into this and trying to repro on our end, hence the "needs-more-info" label.

**Comment by eysi09** [comment]
Hi @carlosjgp 

A few things to unpack here.

> Garden is building the images locally on my laptop instead of spinning up Kaniko pods on the cluster.

Just to confirm, do you mean that you've fallen back to building locally because of the issue with Kaniko described above?

> I've also noticed that Garden is now deploying a new POD into my NS called garden-util-<RS hash>-<POD hash> ... This is meant to run on garden-system

This is actually to be expected. We're phasing out the `garden-system` namespace and instead opting to de-centralize things via individual namespaces. This particular pod is a utility pod that replaces the previous build sync pods. That is, the code gets synced from your computer into this pod, which shares a volume with Kaniko (which eventuall does the building).

So the above looks normal.

> every time I run garden deploy <module> it takes about 20m even without changes

This however does not look normal. Can you try running `garden deploy <service> --show-timestamps --log-level silly`. It will be a lot of output but hopefully helps us diagnose the underlying issue.

**Comment by carlosjgp** [comment]
@eysi09 here is a gist with the logs (obfuscated due to company rules)

```
$ garden version
0.12.23
```

Total time: 20m 3s
At least 1/8 Cores of my laptop running @ 100% on occasions two and up to 3 of them on some peaks

https://gist.github.com/carlosjgp/a479026a432590c806e088afecebf3aa

**Comment by edvald** [comment]
There's a strange number of `Calling configure handler ...` log lines here. I suspect the high CPU usage may have something to do with that, and not the actual builds (which still appear to be happening in Kaniko as expected). So basically I think we're looking at two entirely separate issues here.

@carlosjgp Just to confirm that, could you also share the "silly" logs for the latest Garden version that does work for you?

**Comment by gsvachulay01** [comment]
Same here, I have this going with:
```
Failed building magic. Here is the output:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Unable to query registry for image status: time="2021-08-23T21:43:08Z" level=fatal msg="Error parsing image name \"docker://127.0.0.1:5000/xyz/magic:v-469c9407a8\": Error reading manifest v-469c9407a8 in 127.0.0.1:5000/xyz/magic: received unexpected HTTP
status: 500 Internal Server Error"
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```
on MacOS 10.15.7 Catalina (intel)
using garden 0.12.25 (today's latest, but same with previous too)
on config
```
...
  - name: kubernetes
    context: localk3s
    environments:
    - mac-k3s
    buildMode: kaniko
    # buildMode: cluster-buildkit
    # kaniko:
    #   namespace: null
    deploymentRegistry: null
    # deploymentRegistry:
    #   hostname: garden-docker-registry.garden-system.svc.cluster.local
    #   port: 5000
    #   namespace: library
    setupIngressController: false
    imagePullSecrets:
    - name: registrycreds
      namespace: default
...
```
as you can see from the commented out sections I tried various different combinations and directly addressing the internal registry too. That doesn't work as it's not listening on HTTPS. For every other combination I get the above error. Same is true for docker-buildkit (regardless of rootless or not).

This is extremely disruptive and I can't downgrade to 0.12.21 as the package for MacOS for that release is corrupted and it doesn't execute.

I can positively say using different combinations of `deploymentRegistry: null` or `deploymentRegistry:` makes no difference either.

I'll try setting up a dummy registry for circumventing remote, but some advice or fix would be welcome.

**Comment by edvald** [comment]
> This is extremely disruptive ...

Understood @gsvachulay01! Happy to take a look, although I might need more info.

> ... the package for MacOS for that release is corrupted and it doesn't execute.

That's strange, first time we hear of that. Can you share what comes up when you try it?

> I can positively say using different combinations of deploymentRegistry: null or deploymentRegistry: makes no difference either.

Yeah, the latter as you described it above is definitely not going to work, but omitting `deploymentRegistry` should work without a 500 error for sure. It actually appears these may be different issues, since a 500 error indicates the in-cluster registry is not working at all.

My first suspicion is something specific to k3s that we haven't tested. To my knowledge we've not explicitly validated the in-cluster registry and in-cluster builds to work on k3s, but I'm also not aware of reasons it shouldn't work, especially if it did work before. 

How exactly are you running k3s? I could try the same and see what I find.

I'm also assuming you've tried with a fresh cluster? Or after running `garden plugins kubernetes uninstall-garden-services` on the cluster.

**Comment by edvald** [comment]
Btw @carlosjgp I'm _pretty_ sure your issue is resolved now. We've got probably three separate issues on this thread :P Could you maybe give today's release a spin?

**Comment by gsvachulay01** [comment]
Hi @edvald , happy to explain further. The package is `garden-0.12.21-macos-amd64.tar.gz` and it extracts to the regular directory tree that should go under `garden/bin`.  When opening the garden executable, you get `“garden” is damaged and can’t be opened. You should move it to the Bin.`

For the deploymentRegistry, I tried omitting it completely, setting it to `null`, leaving nothing after the `:` and neither of those yielded a different result.

For k3s, this is just to emulate a linux setup. I've done some customisation of my own to make it as barebones as I possibly could. I removed the traefik ingress controller and anything that I can see would interfere with garden. I use a `setupIngressController: false` setting as garden's nginx doesn't run with `hostNetwork: true` which results in it always trying to create nodePorts for the IngressController instead of using 80 and 443 as the host's own ports. However that issue is solved by switching to vanilla nginx ingress controller and garden picked that up just fine.

I'm running the k3s as a VM inside VirtualBox on the Mac with a reasonable amount of resources. I can connect to it from the mac using both kubectl and other things due to VBox's NAT. I NAT 80 and 443 for the ingress controller and those work fine.

I also have docker-desktop running on the Mac directly, with k8s disabled on it and the docker-desktop context wiped from all config files. I've tried this with a fresh k3s install 3 times, and have after each new provider config change done a complete `delete environment` and `uninstall-garden-services` then `plugins kubernetes cluster-init`. I might try running garden directly from inside the VM, but it would be nice to see it can act as a remote provider as that's the main use case. I also have the provider for k3s set to `kubernetes` not `local-kubernetes` for clarity.

A weird thing I've seen when browsing the k8s api after the plugins were initialised is that the garden-registry has a service listening on 5000 as a ClusterIP and is accessible from other service, but the envoy proxy that came with it has no service, ingress or any other means to be talked to, so I'm not quite sure what that's meant to do.

**Comment by edvald** [comment]
> is damaged and can’t be opened. You should move it to the Bin

Ah I should have realized. This is macOS's security feature being a bit aggressive. I've found it doesn't come up at all if you download (with e.g. wget or curl) and extract via the command line. We're working on getting the builds properly notarized to avoid that for once and all, and our latest release actually has a `garden self-update` command that takes a version argument. I suppose you could use that actually, you'll just lose the `self-update` command and need to upgrade it using brew later.

Just so I understand, did your setup work _for you_ at some point with an older Garden version? Because I'm ~90% sure your issue isn't the same as the others on the thread.

> For the deploymentRegistry, I tried omitting it completely, setting it to null, leaving nothing after the : and neither of those yielded a different result.

Those should indeed all resolve to the same result, which is to deploy an in-cluster registry instead of relying on an external registry. The `deploymentRegistry` is meant to configure an external registry, which is what we generally recommend for remote clusters, especially shared ones.

As for the k3s setup, that's one I have certainly not tried. Usually for local setups, especially if you anyway have docker running locally already, I'd suggest using the local docker for builds. But having k3s in a separate VM will make that a little thorny, so I think in general you have the right idea, you're just a touch off the beaten path with the setup.

One thing worth checking is the logs for `Deployment/garden-docker-registry` in the `garden-system` namespace. Also `DaemonSet/garden-registry-proxy`. Something is returning 500s, and those logs might shine a light. Could you run a build and then check those to see if something relevant comes up?

**Comment by gsvachulay01** [comment]
@edvald you solved it! I was trying a couple other options, but after a second look at the logs during a debug logging session with the CLI, it returned permission errors for PUT calls to the registry. I've exec'd into the registry container and verified that it indeed had no permissions to write `/var/lib/registry/`. I checked the mapped hostPath in the VM and it was the default `root:root` `0655`.

I think this means you most likely need to run a privileged sidecar to fix permissions for the registry as my elasticsearch container that has that, worked fine without modding the FS. After a quick `777` and a pod delete, it works. I now have other bugs, but they appear fixable.

I'm only guessing, but potentially the reason why you don't have this issue elsewhere yet is because it's rare that the engine would not have full control on the path it makes for storage. However, in containerd's recent changes, this is not unusual anymore since the fs and container namespaces have been the cornerstone for rootless access isolation (so this may be more and more prominent soon). Docker and most other lite-k8s deployment would not pop this error either as most of the time they execute as privileged, root or both.

I've now hit build issues with public name resolution, which I suspect are k3s-specific, so I hope this is the end. Please let me know if the fs permission fix will be brought on as a fix.

**Comment by edvald** [comment]
Aha, I see. I can't say yet that I fully understand _why_ this would be the case, but that's good to know. I wouldn't expect (or like) the registry container to require privileged access. 

Maybe it's rather about the registry chart used being outdated and needing to be set up a bit differently, using a different path for data or something... 

_checks the version..._

Yep, we've indeed been using a now-deprecated Helm chart, not tested against recent k8s versions. I don't know if other versions are out that would address this but that's certainly a good issue to flag, thanks for hunting this down with us :)

**Comment by thsig** [comment]
Looks like this was at least partially addressed in #2619, so i'm closing this for now (since we haven't heard any further reports of this).
